### PR TITLE
[feature-threads] Added unlocking spec in root iterator function - MOD-4883

### PR DESCRIPTION
**Pull request by meiravgri**
* Release spec's lock in root rp and in the pager if it declares EOF.
* During query execution, the highlighter searches fields in the spec cache instead of the spec itself

### CODE REVIEW

File: src/result_processor.c [hunk 1]
```diff
@@ -53,9 +53,15 @@ void SearchResult_Destroy(SearchResult *r) {
   SearchResult_Clear(r);
   RLookupRow_Cleanup(&r->rowdata);
 }
-
-static int RPGeneric_NextEOF(ResultProcessor *rp, SearchResult *res) {
-  return RS_RESULT_EOF;
```
Comment (meiravgri) [inline src/result_processor.c:58]: not in use

Comment (GuyAv46) [reply src/result_processor.c:58]: interesting

File: src/spec.c [hunk 2]
```diff
@@ -1643,7 +1643,7 @@ size_t IndexSpec_GetVersion(const IndexSpec *sp) {
 }
 
 // Update the spec vesrion if we update the index.
-// This function should be called after the write lock is acquired.
+// It is assumed that this function is called after locking Redis.
```
Comment (GuyAv46) [inline src/spec.c:1646]: why this change?

File: src/result_processor.h [hunk 3]
```diff
@@ -190,8 +190,10 @@ typedef struct ResultProcessor {
   void (*Free)(struct ResultProcessor *self);
 } ResultProcessor;
 
+// Get the index search context from the result processor
+#define RP_SCTX(rpctx) ((rpctx)->parent->sctx)
 // Get the index spec from the result processor
-#define RP_SPEC(rpctx) ((rpctx)->parent->sctx->spec)
+#define RP_SPEC(rpctx) (RP_SCTX(rpctx)->spec)
```
Comment (GuyAv46) [inline src/result_processor.h:196]: I see that the highlighter uses this macro, so it uses the spec after we unlock it. isn't that a problem?

**Review by GuyAv46 — COMMENTED** [review]
Nice and clean, one main question (please reply in the relevant comment):
Can we actually unlock the spec safely?

File: src/result_processor.h (hunk 3 above)
Comment (meiravgri) [reply src/result_processor.h:196]: Looks like it is used only to search and check if specific fields exist in the spec and that their type is INDEXFLD_T_FULLTEXT
As long as the GC can't remove/modify/move fields from the spec, looks OK

Comment (meiravgri) [reply src/result_processor.h:196]: In second thought, the highlighter doesn't hold the GIL too so no protection at this point on the spec's fields.

File: src/result_processor.c [hunk 4]
```diff
@@ -53,9 +53,15 @@ void SearchResult_Destroy(SearchResult *r) {
...
+static int EndIteratorResult_to_RPResult(int rc) {
```
Comment (GuyAv46) [inline src/result_processor.c:56]:
```suggestion
static int IteratorResult_to_RPResult(int rc) {
```
Or just free the spec in both cases

File: src/spec.c (hunk 2 above)
Comment (meiravgri) [reply src/spec.c:1646]: It was inaccurate in the first place

File: src/result_processor.c [hunk 5]
```diff
@@ -71,6 +62,16 @@ static int EndIteratorResult_to_RPResult(int rc) {
  * downstream.
  *******************************************************************************************************************/
 
+// Get the index search context from the result processor
+#define RP_SCTX(rpctx) ((rpctx)->parent->sctx)
+// Get the index spec from the result processor - this should be used only if the spec 
+// can be accessed safely.
+#define RP_SPEC(rpctx) (RP_SCTX(rpctx)->spec)
+
+static int UnlockSpec_and_ReturnRPResult(ResultProcessor *base, int result_status) {
```
Comment (GuyAv46) [inline src/result_processor.c:71]: this is an enum, lets return the enum type (and get it to the function)

File: src/result_processor.h [hunk 6]
```diff
@@ -154,7 +154,10 @@ typedef enum {
 typedef enum {
   RESULT_PROCESSOR_F_ACCESS_REDIS = 0x01,  // The result processor requires access to redis keyspace.
 
-  RESULT_PROCESSOR_F_BREAKS_PIPELINE = 0x02 // The result processor might break the pipeline by changing RPStatus.
+  // The result processor might break the pipeline by changing RPStatus.
+  // Note that this kind of rp is also responsible to release the spec lock when it breaks the pipeline
+  // (declaring EOF or TIMEOUT).
+  RESULT_PROCESSOR_F_BREAKS_PIPELINE = 0x02
```
Comment (GuyAv46) [inline src/result_processor.h:160]: consider mentioning the `UnlockSpec_and_ReturnRPResult` function

**Review by GuyAv46 — COMMENTED** [review]
Very nice! to small comments

**Review by GuyAv46 — APPROVED** [review]

**Review by DvirDukhan — APPROVED** [review]
